In [90]:
import pandas as pd
import numpy as np
import seaborn as sns

#TO Encode, Scale and split data
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split

#The Models we are going to use
from sklearn.neighbors import KNeighborsClassifier as KNN
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier

#To make a print_score function
from sklearn.model_selection import cross_val_predict, cross_val_score
from sklearn.metrics import confusion_matrix,classification_report,accuracy_score


In [91]:
# Load the data and preprocess it
dataset =  pd.read_csv("agaricus-lepiota.csv")


In [92]:
#dataset.head()
#dataset.info() #Shows that every row has non-null values

In [ ]:
#Wanted to see a heatmap of the correletaiom between the different values and collum.
#This will show the  most indicative features in the dataset and the most domentating feutures. 
corr = dataset.corr()
sns.heatmap(corr)

In [93]:
#Need to drop some coloms because of the logical rules in the dataset.
#This is features that are the most indicative, and we would therefore drop these before running the models. 
dataset = dataset.drop(columns='p.1') #Odor
dataset = dataset.drop(columns='k.1') #Spore-print-color
dataset = dataset.drop(columns='u') #Habitant
dataset = dataset.drop(columns='s.3') #Population
dataset = dataset.drop(columns='s.1')#Above
dataset = dataset.drop(columns='s.2')#below
dataset = dataset.drop(columns='p.3')#ring-type
dataset = dataset.drop(columns='e.1')#stalk-root
dataset = dataset.drop(columns='k')#gill-color
dataset = dataset.drop(columns='t')#bruises?
dataset = dataset.drop(columns='c')#gill-spacing
dataset = dataset.drop(columns='n.1')#gill-size
dataset = dataset.drop(columns='e')#stalk-shape
dataset = dataset.drop(columns='n')#cap-color

In [94]:
#Label Encoding, so that the cells contain corresponding number for the character, and replaces it. 
Encoder = LabelEncoder() 
for col in dataset.columns:
    dataset[col] = Encoder.fit_transform(dataset[col])

In [95]:
# 'p'/class is the column that decides if the mushroom is edible or poisounus.  
X=dataset.drop('p',axis=1) #Predictors 'p'="class", This is what we are trying to predict.
y=dataset['p'] #Response, the data we have to work with, is then the rest of the dataset

In [96]:
scalar = StandardScaler()
X = pd.DataFrame(scalar.fit_transform(X), columns = X.columns)

In [97]:
#Splitting the dataset into the Training set and Test set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=38)

In [98]:
#We want to use KNN, DecisionTreeclassifier and RandomForrestClassifier because 

In [99]:
#KNN
knn = KNN()
knn.fit(X_train,y_train)

#DecisionTreeClassifier
dtc = DecisionTreeClassifier(max_depth=5, criterion="entropy")
dtc.fit(X_train,y_train)

#RandomForrestClassifier
rfc = RandomForestClassifier(oob_score=True)
rfc.fit(X_train, y_train)

RandomForestClassifier(oob_score=True)

In [100]:
#A function to print a classification report for the different models.
def print_score(classifier,X_train,y_train,X_test,y_test, name):
    print("Training results for", name, ":\n")
    print('Classification Report:\n{}\n'.format(classification_report(y_train,classifier.predict(X_train))))

In [101]:
print_score(knn,X_train,y_train,X_test,y_test,"KNN")
print_score(dtc,X_train,y_train,X_test,y_test,"DTC")
print_score(rfc,X_train,y_train,X_test,y_test,"RFC")

Training results for KNN :

Classification Report:
              precision    recall  f1-score   support

           0       0.77      0.90      0.83      2945
           1       0.87      0.72      0.79      2741

    accuracy                           0.81      5686
   macro avg       0.82      0.81      0.81      5686
weighted avg       0.82      0.81      0.81      5686


Training results for DTC :

Classification Report:
              precision    recall  f1-score   support

           0       0.94      0.52      0.67      2945
           1       0.65      0.96      0.78      2741

    accuracy                           0.73      5686
   macro avg       0.79      0.74      0.72      5686
weighted avg       0.80      0.73      0.72      5686


Training results for RFC :

Classification Report:
              precision    recall  f1-score   support

           0       0.88      0.79      0.83      2945
           1       0.80      0.88      0.84      2741

    accuracy               

c) Our best performing model is the RandomForrest classifier, as the average score is 83%. We would still not trust the model as there is still 20% that is wrong, and that can be dangerous. The dataset was originally sufficient because it would give a score of 100% for all three models. To really test the models we would need to decrease the numbers of colums and avoid the logical rules of the dataset. 

The decision tree classifier is a tree-like model that splits the data based on the most significant feature until a prediction can be made.
We think the random forrest classifier perform best because it is an ensemble of decision trees, where each tree makes a prediction and the final prediction is made by combining the predictions of all trees. Therefore, the RFC is even more precise than the dtc. 